### PR TITLE
C2C-3 Review and fix Rachio Integration

### PR DESCRIPTION
**Pull request by marstorp**
various exception fixes, removal of capability polling
removal of unnecessary null-point checks
indentations now all spaces

### CODE REVIEW

File: devicetypes/rachio/rachio-iro2-controller.src/rachio-iro2-controller.groovy [hunk 1]
```diff
@@ -158,79 +157,84 @@ metadata {
...
+//    def gen = state.deviceId ? parent?.getDevGeneration(state.deviceId) : null
+//    updateDataValue("model", "${device.name}${gen ? " ($gen)" : ""}")
 }
 
 void installed() {
-	initialize()
-	state.isInstalled = true
+    initialize()
+    state.isInstalled = true
+    sendEvent(name: "DeviceWatch-DeviceStatus", value: "offline", displayed: false, isStateChange: true)
```
Comment (workingmonk) [inline devicetypes/rachio/rachio-iro2-controller.src/rachio-iro2-controller.groovy]: change to online

File: devicetypes/rachio/rachio-iro2-controller.src/rachio-iro2-controller.groovy [hunk 2]
```diff
@@ -334,77 +325,57 @@ def hardwareModelEvent(val) {
...
 }
 
 def setZoneWaterTime(timeVal) {
-    def curState = device?.currentState("curZoneWaterTime")?.value.toString()
     def newVal = timeVal ? timeVal.toInteger() : parent?.settings?.defaultZoneTime.toInteger()
-    if(isStateChange(device, "curZoneWaterTime", newVal.toString())) {
-        log.debug("UPDATED: Manual Zone Water Time (${newVal}) | Previous: (${curState})")
-        sendEvent(name: 'curZoneWaterTime', value: newVal, displayed: true)
-    }
+    log.debug("Manual Zone Water Time (${newVal}) | Previous: (${curState})")
```
Comment (workingmonk) [inline devicetypes/rachio/rachio-iro2-controller.src/rachio-iro2-controller.groovy]: update log statement to remove curState

File: devicetypes/rachio/rachio-iro2-controller.src/rachio-iro2-controller.groovy [hunk 3]
```diff
@@ -420,38 +391,28 @@ def scheduleDataEvent(sData, zData, rainDelay) {
...
+    sendEvent(name: 'curZoneName', value: zoneName?.toString(), displayed: true)
+
+    log.debug("Active Zone Number (${zoneNum})")
+    sendEvent(name: 'curZoneNumber', value: zoneNum, displayed: true)
+    log.debug("Zone Cycle Count (${zoneCycleCount})")
+    sendEvent(name: 'curZoneCycleCount', value: zoneCycleCount, displayed: true)
+
+    sendEvent(name: 'curZoneIsCycling', value: zoneIsCycling?.toString().capitalize(), displayed: true)
+
+    log.debug("Zone StartDate (${(zoneStartDate ? epochToDt(zoneStartDate).toString() : "Not Active")})")
```
Comment (workingmonk) [inline devicetypes/rachio/rachio-iro2-controller.src/rachio-iro2-controller.groovy]: null safe toString()

File: devicetypes/rachio/rachio-iro2-controller.src/rachio-iro2-controller.groovy [hunk 4]
```diff
@@ -477,145 +438,145 @@ def setRainDelayString( rainDelay) {
...
 }
 
 def doSetRainDelay() {
     def value = device.latestValue('rainDelay')
-    log.debug("Set Rain Delay ${value}")
-    def res = parent?.setRainDelay(this, state?.deviceId, value);
-    if( !res) {
+    log.debug "Set Rain Delay ${value}"
+    if (parent?.setRainDelay(this, state.deviceId, value)) {
+        sendEvent(name: 'rainDelay', value: value)
```
Comment (workingmonk) [inline devicetypes/rachio/rachio-iro2-controller.src/rachio-iro2-controller.groovy]: remove this sendEvent

File: devicetypes/rachio/rachio-iro2-controller.src/rachio-iro2-controller.groovy [hunk 5]
```diff
@@ -477,145 +438,145 @@ def setRainDelayString( rainDelay) {
...
+        close()
+    } else {
+        log.debug "Switch is Already OFF... Ignoring..."
+    }
 }
 
 def open() {
-    log.trace "open()..."
-    log.info "open command is not currently supported by the controller device..."
+    log.debug "open command is not currently supported by the controller device..."
```
Comment (workingmonk) [inline devicetypes/rachio/rachio-iro2-controller.src/rachio-iro2-controller.groovy:561]: this should call runAllZones

File: devicetypes/rachio/rachio-iro2-zone.src/rachio-iro2-zone.groovy [hunk 6]
```diff
@@ -13,21 +13,20 @@
  *  on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License
  *  for the specific language governing permissions and limitations under the License.
  *
- *	Modified: 7-30-2018
+ *    Modified: 8-13-2018
```
Comment (workingmonk) [inline devicetypes/rachio/rachio-iro2-zone.src/rachio-iro2-zone.groovy]: remove modified. Git is keeping track of that.

File: devicetypes/rachio/rachio-iro2-zone.src/rachio-iro2-zone.groovy [hunk 7]
```diff
@@ -173,86 +173,88 @@ def parse(String description) {
...
-    updateDataValue("model", "${device?.name}${gen ? " ($gen)" : ""}")
+// getDevGeneration is not defined in the connect app...
+//    def gen = state.deviceId ? parent?.getDevGeneration(state.deviceId) : null
+//    updateDataValue("model", "${device?.name}${gen ? " ($gen)" : ""}")
 }
 
 void installed() {
     state.isInstalled = true
     initialize()
+    sendEvent(name: "DeviceWatch-DeviceStatus", value: "offline", displayed: false, isStateChange: true)
```
Comment (workingmonk) [inline devicetypes/rachio/rachio-iro2-zone.src/rachio-iro2-zone.groovy]: should be online

File: smartapps/rachio/rachio-connect.src/rachio-connect.groovy [hunk 8]
```diff
@@ -968,19 +973,19 @@ def pollChildren(childDev, devData) {
...
         if(childDev && devData) {
-            String dni = childDev?.device?.deviceNetworkId
-            String devLabel = childDev?.label
+            String dni = childDev.device?.deviceNetworkId
+            String devLabel = childDev.label
             def schedData = devData.currentSchedule
             def devStatus = devData
             def rainDelay = getCurrentRainDelay(devStatus)
             def status = devStatus?.status
             def onlStatus = status?.toString()?.toLowerCase() == "online" ? "online" : "offline"
```
Comment (workingmonk) [inline smartapps/rachio/rachio-connect.src/rachio-connect.groovy]: removed onlStatus

File: smartapps/rachio/rachio-connect.src/rachio-connect.groovy [hunk 9]
```diff
@@ -1281,11 +1282,15 @@ def getControlLblById(id) {
...
-	// convert to configured rain delay to days.
-	def ret =  (res?.rainDelayExpirationDate || res?.rainDelayStartDate) ? (res?.rainDelayExpirationDate - res?.rainDelayStartDate)/(26*60*60*1000) : 0
-	def value = (long) Math.floor(ret + 0.5d)
-	return value
+    //log.debug("getCurrentRainDelay($devId)...")
+    // convert to configured rain delay to days.
+    // as it seems rainDelayStartDate is never returned in the API response and it seems the value doesn't change using hardcoded default value of
+    // 1534193839000 as (1534287439000 - 1534193839000)/(26*60*60*1000) = 1
+    //def ret =  (res?.rainDelayExpirationDate || res?.rainDelayStartDate) ? (res?.rainDelayExpirationDate - res?.rainDelayStartDate)/(26*60*60*1000) : 0
+    def rainDelayStartDate = 1534193839000
```
Comment (workingmonk) [inline smartapps/rachio/rachio-connect.src/rachio-connect.groovy]: `def rainDelayStartDate = res?.rainDelayStartDate ?: 1534193839000`

File: devicetypes/rachio/rachio-iro2-controller.src/rachio-iro2-controller.groovy (hunk 5 above)
Comment (marstorp) [reply devicetypes/rachio/rachio-iro2-controller.src/rachio-iro2-controller.groovy:561]: No as the developer at Rachio did this

File: smartapps/rachio/rachio-connect.src/rachio-connect.groovy (hunk 9 above)
Comment (marstorp) [reply smartapps/rachio/rachio-connect.src/rachio-connect.groovy]: set rainDelayStartDate to current date

**Review by workingmonk — APPROVED** [review]

**Comment by marstorp** [comment]
LGTM ⛵️

**Pull request merged**